File: packages/react-native-vision-camera/ios/Hybrid Objects/Constraints/HybridCameraSessionConfig.swift
///
/// HybridCameraSessionConfig.swift
/// VisionCamera
/// Copyright © 2025 Marc Rousavy @ Margelo
///

import AVFoundation
import Foundation
import NitroModules

final class HybridCameraSessionConfig: HybridCameraSessionConfigSpec, NativeCameraSessionConfig {
  let negotiatedFormat: NegotiatedFormat
  let enabledConstraints: EnabledConstraints

  init(negotiatedFormat: NegotiatedFormat, enabledConstraints: EnabledConstraints) {
    self.negotiatedFormat = negotiatedFormat
    self.enabledConstraints = enabledConstraints
  }

  var selectedFPS: Double? {
    return enabledConstraints.selectedFPS
  }

  var selectedVideoStabilizationMode: TargetStabilizationMode? {
    return enabledConstraints.selectedVideoStabilizationMode
  }

  var selectedPreviewStabilizationMode: TargetStabilizationMode? {
    return enabledConstraints.selectedPreviewStabilizationMode
  }

  var selectedVideoDynamicRange: TargetDynamicRange? {
    return enabledConstraints.selectedVideoDynamicRange
  }

  var isPhotoHDREnabled: Bool {
    // Highest Photo Quality implies Smart HDR, this is not directly controllable.
    return negotiatedFormat.format.isHighestPhotoQualitySupported
  }

  var nativePixelFormat: PixelFormat {
    let format = negotiatedFormat.format
    return PixelFormat(mediaSubType: format.formatDescription.mediaSubType)
  }

  var autoFocusSystem: AutoFocusSystem {
    let format = negotiatedFormat.format
    return AutoFocusSystem(af: format.autoFocusSystem)
  }

  var isBinned: Bool {
    return negotiatedFormat.format.isVideoBinned
  }

  func toString() -> String {
    let components = [
      "format: \(negotiatedFormat.format)",
      "selectedFPS: \(String(describing: selectedFPS))",
      "selectedVideoStabilizationMode: \(String(describing: selectedVideoStabilizationMode))",
      "selectedPreviewStabilizationMode: \(String(describing: selectedPreviewStabilizationMode))",
      "selectedVideoDynamicRange: \(String(describing: selectedVideoDynamicRange))",
      "isPhotoHDREnabled: \(isPhotoHDREnabled)",
    ]
    return "CameraSessionConfig(\(components.joined(separator: ", ")))"
  }
}

File: packages/react-native-vision-camera/android/src/main/java/com/margelo/nitro/camera/hybrids/HybridCameraSessionConfig.kt
package com.margelo.nitro.camera.hybrids

import androidx.camera.core.CameraInfo
import androidx.camera.core.SessionConfig
import com.margelo.nitro.camera.AutoFocusSystem
import com.margelo.nitro.camera.HybridCameraSessionConfigSpec
import com.margelo.nitro.camera.PixelFormat
import com.margelo.nitro.camera.TargetDynamicRange
import com.margelo.nitro.camera.TargetStabilizationMode
import com.margelo.nitro.camera.public.NativeCameraOutput
import com.margelo.nitro.camera.public.NativeCameraSessionConfig

class HybridCameraSessionConfig(
  private val cameraInfo: CameraInfo,
  override val sessionConfig: SessionConfig,
  private val resolvedConfig: NativeCameraOutput.Config,
) : HybridCameraSessionConfigSpec(),
  NativeCameraSessionConfig {
  override val selectedFPS: Double?
    get() = resolvedConfig.fpsRange?.upper?.toDouble()

  override val selectedVideoStabilizationMode: TargetStabilizationMode?
    get() = resolvedConfig.videoStabilizationMode

  override val selectedPreviewStabilizationMode: TargetStabilizationMode?
    get() = resolvedConfig.previewStabilizationMode

  override val selectedVideoDynamicRange: TargetDynamicRange?
    get() = resolvedConfig.videoDynamicRange

  override val isPhotoHDREnabled: Boolean
    get() = resolvedConfig.photoHDR ?: false

  override val nativePixelFormat: PixelFormat
    // TODO: Do we always stream in PRIVATE?
    get() = PixelFormat.PRIVATE

  override val autoFocusSystem: AutoFocusSystem
    // TODO: Is PHASE_DETECTION correct on Android? Should we expose UNKNOWN instead?
    get() = AutoFocusSystem.PHASE_DETECTION

  override val isBinned: Boolean
    // In CameraX, currently every stream is binned - see https://issuetracker.google.com/issues/500394299
    get() = true

  override fun toString(): String {
    val components =
      arrayOf(
        "config: $sessionConfig",
        "selectedFPS: $selectedFPS",
        "selectedVideoStabilizationMode: $selectedVideoStabilizationMode",
        "selectedPreviewStabilizationMode: $selectedPreviewStabilizationMode",
        "selectedVideoDynamicRange: $selectedVideoDynamicRange",
        "isPhotoHDREnabled: $isPhotoHDREnabled",
      )
    return "CameraSessionConfig(${components.joinToString(", ")})"
  }
}
